tests/integration.rs

```diff
@@ -0,0 +1,495 @@
+    OrderStatus,
+};
+use rcgen::{CertificateParams, DistinguishedName, KeyPair};
+use rustls::client::{verify_server_cert_signed_by_trust_anchor, verify_server_name};
+use rustls::crypto::CryptoProvider;
+use rustls::pki_types::pem::PemObject;
+use rustls::pki_types::{CertificateDer, ServerName};
+use rustls::server::ParsedCertificate;
+use rustls::RootCertStore;
+use rustls_pki_types::UnixTime;
+use serde::ser::SerializeMap;
+use serde::{Serialize, Serializer};
+use tempfile::NamedTempFile;
+use tokio::net::TcpStream;
+use tokio::time::sleep;
+use tracing::{debug, error, info, trace};
+use tracing_subscriber::layer::SubscriberExt;
+use tracing_subscriber::util::SubscriberInitExt;
+use tracing_subscriber::{fmt, EnvFilter};
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn StdError>> {
+    tracing_subscriber::registry()
+        .with(fmt::layer())
+        .with(EnvFilter::from_default_env())
+        .init();
+
+    // Spawn a Pebble CA and a challenge response server.
+    debug!("starting Pebble CA environment");
+    let pebble = PebbleEnvironment::new(DEFAULT_CONFIG.clone()).await?;
+
+    // Create a test account with the Pebble CA.
+    debug!("creating test account");
+    let (mut account, _) = Account::create_with_http(
+        &NewAccount {
+            contact: &[],
+            terms_of_service_agreed: true,
+            only_return_existing: false,
+        },
+        &pebble.directory_url(),
+        None,
+        Box::new(pebble.http_client()),
+    )
+    .await?;
+    info!("created ACME account: {}", account.id());
+
+    // Issue a certificate w/ HTTP-01 challenge and verify it.
+    pebble.test_http1(&mut account).await?;
+
+    Ok(())
+}
+
+/// A test environment running Pebble and a challenge test server.
+///
+/// Subprocesses are torn down cleanly on drop to avoid leaving
+/// stray child processes.
+struct PebbleEnvironment {
+    config: Config,
+    #[allow(dead_code)] // Held for the lifetime of the environment.
+    config_file: NamedTempFile,
+    #[allow(dead_code)] // Held for the lifetime of the environment.
+    pebble: Subprocess,
+    #[allow(dead_code)] // Held for the lifetime of the environment.
+    challtestsrv: Subprocess,
+}
+
+impl PebbleEnvironment {
+    /// Create a test environment for the given configuration.
+    ///
+    /// Set the PEBBLE and CHALLTESTSRV to pebble and pebble-challtestsrv binaries
+    /// respectively. If unset "./pebble" and "./pebble-challtestsrv" are used.
+    ///
+    /// Returns only once the Pebble CA server interface is responding.
+    async fn new(config: Config) -> io::Result<Self> {
+        let config_file = NamedTempFile::new()?;
+        let config_json = serde_json::to_string_pretty(&config)?;
+        trace!("pebble config:\n{config_json}\n");
+        fs::write(&config_file, config_json)?;
+
+        let pebble_path = env::var("PEBBLE").ok().unwrap_or("./pebble".to_owned());
```

Comment:
Nit: use `unwrap_or_else()` here?

Reply:
Switched from `ok().unwrap_or()` to just `unwrap_or()`, good catch. Using `unwrap_or_else()` introduces an unnecessary closure.

Reply:
So I did actually mean `unwrap_or_else()`, so the allocation from `"./pebble".to_owned()` is only done when necessary. See https://rust-lang.github.io/rust-clippy/master/index.html#/unwrap_or_else.

Reply:
Ah, fair enough. Switched.